chain/network/src/peer/peer_actor.rs

```diff
@@ -807,46 +697,38 @@ impl Actor for PeerActor {
 
     fn stopped(&mut self, _ctx: &mut Self::Context) {
         actix::Arbiter::current().stop();
-        let metric_label = &self.peer_addr.to_string();
-        // Garbage collect our metrics, ignoring the case where the metric is
-        // already deleted, to avoid races when a peer tries to reconnect.
-        let _ = metrics::PEER_DATA_READ_BUFFER_SIZE.remove_label_values(&[metric_label]);
-        let _ = metrics::PEER_DATA_WRITE_BUFFER_SIZE.remove_label_values(&[metric_label]);
     }
 }
 
-impl WriteHandler<io::Error> for PeerActor {}
-
-#[derive(actix::Message)]
-#[rtype(result = "()")]
-struct RecvMessage(Result<Vec<u8>, RecvError>);
+impl actix::Handler<stream::Error> for PeerActor {
+    type Result = ();
+    fn handle(&mut self, err: stream::Error, ctx: &mut Self::Context) {
+        let expected = match &err {
+            stream::Error::Recv(stream::RecvError::MessageTooLarge { .. }) => {
+                self.ban_peer(ctx, ReasonForBan::Abusive);
+                true
+            }
+            stream::Error::Send(stream::SendError::QueueOverflow { .. }) => false,
+            stream::Error::Recv(stream::RecvError::IO(err))
+            | stream::Error::Send(stream::SendError::IO(err)) => match err.kind() {
+                io::ErrorKind::UnexpectedEof => true,
+                io::ErrorKind::ConnectionReset => true,
+                _ => false,
+            },
+        };
+        if !expected {
+            error!(target: "network", ?err, "Closing connection to {}", self.peer_info);
+        }
+        ctx.stop();
+    }
+}
 
-impl actix::Handler<RecvMessage> for PeerActor {
+impl actix::Handler<stream::Frame> for PeerActor {
     type Result = ();
     #[perf]
-    fn handle(&mut self, msg: RecvMessage, ctx: &mut Self::Context) {
+    fn handle(&mut self, msg: stream::Frame, ctx: &mut Self::Context) {
```

Comment:
```suggestion
    fn handle(&mut self, stream::Frame(msg): stream::Frame, ctx: &mut Self::Context) {
```

might be a bit nicer (makes it obvious we unpacking everything there is)

Reply:
done